packages/RecentChangesAPI/src/custom/modules/Users/clients/base/api/CustomRecentChangesApi.php

```diff
@@ -0,0 +1,214 @@
+     * @param ServiceBase $api
+     * @param array $args The arguments passed to the GET request. All params are required:
+     *      since: the datetime from which you want to get changes
+     *      modules: a comma separated list of modules that will be checked
+     *      add_deleted: Whether or not "'deleted' = 0" should be added to Where clause of generated SQL query. Value should be 0 or 1
+     * @return array JSON array containing the server's current time and a list of users with changes
+     * @throws SugarApiExceptionNotAuthorized if the user is not an admin
+     */
+    public function getRecentChanges(ServiceBase $api, array $args)
+    {
+        // Ensure the current user is an admin
+        global $current_user;
+        if (!$current_user->is_admin) {
+            throw new SugarApiExceptionNotAuthorized('You do not have admin permissions');
+        }
+
+        // Validate the params
+        $sinceParam = $this->validateSinceParam($args['since']);
+        $AddDeletedParam = $this->validateAddDeletedParam($args['add_deleted']);
+        $modulesParam = $this->validateModulesParam($args['modules']);
+
+        // Get the server's current time
+        // TODO:  Is this the right way to get the current server timestamp?
+        global $timedate;
+        $currentTime = $timedate->getNow()->format($this->dateFormat);
+
+        // Query for updated records
+        $arrayOfQueryResults = $this->queryForRecentChanges($sinceParam, $currentTime, $modulesParam, $AddDeletedParam);
+
+        // Return the formatted results
+        return $this->convertToResponseArray($currentTime, $arrayOfQueryResults);
+    }
+
+    /**
+     * Query for the list of user_ids and associated user_names for users who have records updated between $dateTimeOne
+     * and $dateTimeTwo
+     * @param $dateTimeOne Search for records updated after this dateTime
+     * @param $dateTimeTwo Search for records updated up to and including this dateTime
+     * @param $modules List of modules to query for updates
+     * @param $addDeleted Whether or not 'deleted' = 0 should be added to Where clause of generated SQL query
+     * @return array Array of query results. Each module in $modules will have its own set of results in the array.
+     */
+    protected function queryForRecentChanges($dateTimeOne, $dateTimeTwo, $modules, $addDeleted)
+    {
+        $arrayOfQueryResults = array();
+
+        foreach ($modules as $module) {
+            $query = new SugarQuery();
+            $query->distinct(true);
+            $query->from(BeanFactory::newBean($module), array('team_security' => true, 'add_deleted' => $addDeleted));
+            //TODO:  I was getting invalid link errors for modules like Accounts and Opportunities when I tried to do a
+            // join instead of a joinTable.  Is there a better way to get the table name than lower casing the module name?
+            $tableName = strtolower($module);
```

Comment:
IIRC, with "join()" you are actually specifying the relationship name instead of the module name. Relationship names often follow the convention of being the lowercase name of the module being related to but that's not always the case.

joinTable() lets you specify table by name. Table names are usually the lowercase of module name but it's possible to change this value in Vardefs. There's a parameter on SugarBean called `$table_name` that you could use instead.  So you could try something like `$bean = BeanFactory::newBean($module); $bean->table_name;`

Reply:
Or maybe check the $dictionary global variable. Not sure if one is preferred over other. Generally, using a global would make this harder to unit test.

```
global $dictionary;
$dictionary[$module]['table']
```

Reply:
join() does need the relationship name:
$query->join('tai_accountproducts_accounts', array('alias' => 'apa'));

Back in June of 2015 I found that I had to specify an alias or I could not use the joined table's fields in a where() function:
$query->where()->equals('apa.id',$acct->id);
I'm unsure if this was a bug and if it was, if it has been fixed.

Reply:
Thanks for the info @kuehltha!